Add vue routing to make SPA

andres-holguin opened the pull request:
# Vue Routing
This PR adds vue routing to the application, making it a Single Page Application. This facilitates communication across its pages. In particular it enables the search bar in the landing page to produce a result on the search results page.

![Starfish-Home-Router-V1](https://user-images.githubusercontent.com/44207690/88729710-833a5780-d102-11ea-89f1-054f3cac7c14.gif)

andres-holguin committed:
Create the HomePage view

Renamed from LandingPage.vue

andres-holguin committed:
Move Navbar from SearchPage.vue to index.html

Navbar will now be a part of the main app, not tied to a single view.

andres-holguin committed:
index.html: Implement vue routing

- Add link to VueRouter script.
- Add a router-view which can display Home page or Search page.
- Change how the main app is mounted.

andres-holguin committed:
Tweak HomePage.vue

Tweak to better reflect that mock-search doesn't do anything, yet. Added TODOs.

andres-holguin committed:
Remove MockSearchButton from Home page

andres-holguin committed:
SearcBar.vue: Redefine data initialization

School and course now initialize based on url query params (should be null if DNE), error initializes to null for consistency.

andres-holguin committed:
SearchBar.vue: Abstract out form validation

Create a computed value for whether the form is valid, and methods for how to handle valid and invalid @submit events.

andres-holguin committed:
SearchBar.vue: Handle valid/invalid form

Define the functions for handling valid/invalid forms, and rename queryDatabase to fetchAndEmitResults.

andres-holguin committed:
SearchBar.vue: Modify fetchAndEmitResults()

Modify to set search params from data, and return the promise from the fetch().

andres-holguin committed:
SearchBar.vue: Define triggers for fetchAndEmitResults()

The SearchBar components will now only fetch data from the backend when it mounts or the route changes, and will do so based on the query params.

andres-holguin committed:
HomePage.vue: Remove redirect from image

Removed a router-link on the image that was used for testing the Vue Router.

andres-holguin committed:
HomePage.vue: Remove unused event listener.

Removed an event listener that was never implemented.

Aradhya-B reviewed (approved):
Looks great! Is `HomePage.vue` really the landing page at this point? Could the component made in #85 be put into `HomePage.vue`?

andres-holguin commented:
> Looks great! Is `HomePage.vue` really the landing page at this point? Could the component made in #85 be put into `HomePage.vue`?

That can definitely be done :) #85 can have its HTML converted to a template, and its CSS can be moved into that same component.

keelywan reviewed (approved):
This looks really good! I'll move the HTML in #85 into a Vue component 😄

andres-holguin committed:
SearchBar.vue: Hotfix for query string

The search parameters being set for the fetch function to /search where being set even if the input was null. This fixes that so only non-null parameters get set.

cftorres commented on src/main/webapp/components/SearchBar.vue:
```diff
@@ -27,36 +27,63 @@
       let url = new URL("/search", window.location.href);
-      url.searchParams.set('school', school);
-      url.searchParams.set('course', course);
+      this.school && url.searchParams.set('school', this.school);
```
Perhaps I just don't know about this syntax, but why do you need to use this here when you're just calling a setter?

andres-holguin replied:
I was having an issue where if the school was null, it would set the url search param for school=null, which the back end would read as school = "null" and then try to look for schools named null XD this makes sure null doesn't end up in the query string.

cftorres replied:
oh interesting!

cftorres commented on src/main/webapp/index.html:
```diff
@@ -6,22 +6,46 @@
+
+  // 3. Create the router instance and pass the `routes` option
+  const router = new VueRouter({
+    routes // short for `routes: routes`
```
nit: if you're likely not going to add anything to this object, especially since it's just a shorthand, I think you can include this inline above

andres-holguin replied:
Makes sense, I can make that work :)

cftorres reviewed (approved):
awesome!

nickKelman reviewed (approved):
Awesome change.

andres-holguin committed:
Condense router contstruction with inline routes
